### PR TITLE
tidb-server, terror: add cleanup when `createServer` exits abnormally

### PR DESCRIPTION
**Pull request by zimulala**
## What have you changed? (mandatory)
Handle #6946

## What are the type of the changes (mandatory)?

- Improvement 

## How has this PR been tested (mandatory)?
Before this PR：
```
2018/06/30 09:55:23.368 server.go:190: [warning] Secure connection is NOT ENABLED
2018/06/30 09:55:23.368 terror.go:327: [fatal] listen tcp 0.0.0.0:4000: bind: address already in use
/.../src/github.com/pingcap/tidb/server/server.go:173:
```
After this PR：
```
2018/06/30 09:55:31.783 server.go:190: [warning] Secure connection is NOT ENABLED
2018/06/30 09:55:31.785 delete_range.go:110: [info] [ddl] closing delRange session pool
2018/06/30 09:55:31.785 ddl_worker.go:80: [info] [ddl] close DDL worker 0
2018/06/30 09:55:31.785 ddl.go:382: [info] [ddl] closing DDL:8ab4f703-ed52-43c4-8dff-406b50aa39da takes time 344.571µs
2018/06/30 09:55:31.785 ddl.go:334: [info] [ddl] stop DDL:8ab4f703-ed52-43c4-8dff-406b50aa39da
2018/06/30 09:55:31.787 domain.go:652: [info] [stats] init stats info takes 4.577364ms
2018/06/30 09:55:31.787 domain.go:401: [info] [domain] close
2018/06/30 09:55:31.788 terror.go:335: [fatal] listen tcp 0.0.0.0:4000: bind: address already in use
/.../src/github.com/pingcap/tidb/server/server.go:173:
```

### CODE REVIEW

File: terror/terror.go [hunk 1]
```diff
@@ -328,6 +328,14 @@ func MustNil(err error) {
 	}
 }
 
+// CleanNotNil cleans up and fatals if err is not nil.
```
Comment (shenli) [inline terror/terror.go]: Why put it in the terror package?

**Comment by shenli** [comment]
rest LGTM

File: terror/terror.go [hunk 2]
```diff
@@ -328,6 +328,14 @@ func MustNil(err error) {
...
+func CleanNotNil(err error, closeFun func()) {
```
Comment (coocood) [inline terror/terror.go]: Why not name it `FatalNotNil`?
And if the `closeFun` is optional, then we can use it in more places.

File: terror/terror.go (hunk 1 above)
Comment (coocood) [reply terror/terror.go]: How about we reuse the `MustNil` function, and change it like
```
func MustNil(err error, closeFn ...func())
```

```

File: tidb-server/main.go [hunk 3]
```diff
@@ -487,11 +488,15 @@ func runServer() {
 	}
 }
 
+func close() {
```
Comment (shenli) [inline tidb-server/main.go]: s/close/closeDomainAndStoreage/

**Comment by zimulala** [comment]
PTAL @shenli @coocood

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
Do not forget to attach labels to your PR and cherry-pick this to the release-2.0 branch.

**Comment by shenli** [comment]
@coocood @tiancaiamao  PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**